### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by uvafan**
- Allow for ModificationConstraints which only restrict which words can be modified
- Separate search methods from attacks to be more consistent with our framework
- Update beam/greedy search to be transformation-agnostic
- Add validation for compatibility between constraints and transformations, and search methods and transformations

### CODE REVIEW

File: textattack/search_methods/search_method.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
+from textattack.shared.utils import default_class_repr
+
+class SearchMethod:
+    """
+    This is an abstract class that contains main helper functionality for
```
Comment (jxmorris12) [inline textattack/search_methods/search_method.py:5]: Can you add a sentence about what a search method is?

File: textattack/shared/attack.py [hunk 2]
```diff
@@ -4,29 +4,32 @@
...
 class Attack:
     """
     An attack generates adversarial examples on text. 
     
     This is an abstract class that contains main helper functionality for 
-    attacks. An attack is comprised of a search method, a goal function, and a 
-    transformation, as well as one or more linguistic constraints that 
-    successful examples must meet.
+    attacks. An attack is comprised of a search method, which makes use of
+    a goal function, a transformation, and a set of one or more linguistic
```
Comment (jxmorris12) [inline textattack/shared/attack.py:17]: this is ambiguous language -

maybe say "a search method and corresponding goal function, a transformation, ..."

File: textattack/transformations/composite_transformation.py [hunk 3]
```diff
@@ -16,4 +16,9 @@ def __call__(self, *args, **kwargs):
                 transformation(*args, **kwargs)
             )
         return list(new_tokenized_texts)
-        
\ No newline at end of file
+    
+    def consists_of(self, validator):
```
Comment (jxmorris12) [inline textattack/transformations/composite_transformation.py:20]: add docstring

File: textattack/shared/validators.py [hunk 4]
```diff
@@ -72,4 +72,8 @@ def validate_model_gradient_word_swap_compatibility(model):
     if isinstance(model, textattack.models.helpers.LSTMForClassification):
         return True
     else:
-        raise ValueError(f'Cannot perform GradientBasedWordSwap on model {model}.')
\ No newline at end of file
+        raise ValueError(f'Cannot perform GradientBasedWordSwap on model {model}.')
+
+def is_word_swap(transformation):
```
Comment (jxmorris12) [inline textattack/shared/validators.py:77]: add docstring

File: textattack/shared/attack.py [hunk 5]
```diff
@@ -49,7 +67,7 @@ def get_transformations(self, text, original_text=None,
             transformation: 
             text:
             original text (:obj:`type`, optional): Defaults to None. 
-            apply_constraints:
+            apply_constraints: Whether or not to apply (non-modification) constraints
```
Comment (jxmorris12) [inline textattack/shared/attack.py:70]: nit: remove parentheses

File: textattack/constraints/semantics/repeat_modification.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
+""" Abstract classes represent constraints on text adversarial examples. 
+"""
+
+from textattack.shared.utils import default_class_repr
+from textattack.constraints import ModificationConstraint
+
+class RepeatModification(ModificationConstraint):
+    """ 
+    A constraint disallowing the modification of words which have already been modified
```
Comment (jxmorris12) [inline textattack/constraints/semantics/repeat_modification.py:9]: nit: add period (don't forget these docstring will show up on readthedocs)

File: textattack/constraints/grammaticality/language_models/google_language_model/google_language_model.py [hunk 7]
```diff
@@ -96,7 +100,7 @@ def get_probs(x, x_adv_list):
         # same order they were passed in.
         max_el_indices.sort()
         
-        return np.array(x_adv_list)[max_el_indices]
+        return list(np.array(x_adv_list)[max_el_indices])
```
Comment (jxmorris12) [inline textattack/constraints/grammaticality/language_models/google_language_model/google_language_model.py:103]: Why does this have to be a list?

If it really does have to be a list, then it's not necessarily faster to use the masking of nparrays in this case, and maybe we should just do
```
return [x_adv_list[i] for i in max_el_indices]
```

Maybe the masking is nicer though, and the performance difference should be negligible. up to you.

File: textattack/constraints/constraint.py [hunk 8]
```diff
@@ -20,13 +20,49 @@ def call_many(self, x, x_adv_list, original_text=None, **kwargs):
...
+        incompatible_x_advs = []
+        compatible_x_advs = []
+        for x_adv in x_adv_list:
+            try:
+                if self.check_compatibility(x_adv.attack_attrs['last_transformation']):
+                    compatible_x_advs.append(x_adv)
+                else:
+                    incompatible_x_advs.append(x_adv)
+            except KeyError:
+                raise KeyError('x_adv must have `last_transformation` attack_attr to apply GoogLM constraint')
```
[the remark this answers is not in the document]
Comment (uvafan) [reply textattack/constraints/constraint.py:32]: copy paste oops

File: textattack/constraints/constraint.py [hunk 9]
```diff
@@ -20,13 +20,49 @@ def call_many(self, x, x_adv_list, original_text=None, **kwargs):
...
+        return self._check_constraint(x_adv, original_text=original_text)
+
+    def _check_constraint(self, x, x_adv, original_text=None):
         """ Returns True if C(x,x_adv) is true. """
         raise NotImplementedError()
-    
+
+    def check_compatibility(self, transformation):
+        """ 
+        Checks if this constraint is compatible with the given transformation.
```
[the remark this answers is not in the document]
Comment (uvafan) [reply textattack/constraints/constraint.py:60]: Let me know if the revised version looks OK haha

File: textattack/transformations/transformation.py [hunk 10]
```diff
@@ -7,11 +7,26 @@ class Transformation:
...
+        return transformed_texts
+   
+    def _get_transformations(self, tokenized_text, indices_to_modify):
         raise NotImplementedError()
-    
+
     def extra_repr_keys(self): 
         return []
+
+    def consists_of(self, validator):
```
[the remark this answers is not in the document]
Comment (uvafan) [reply textattack/transformations/transformation.py:29]: removed as discussed

File: textattack/constraints/semantics/sentence_encoders/sentence_encoder.py [hunk 11]
```diff
@@ -93,7 +93,11 @@ def _score_list(self, x, x_adv_list):
             x_list_text = []
             x_adv_list_text = []
             for x_adv in x_adv_list:
-                modified_index = x_adv.attack_attrs['modified_word_index']
+                #@TODO make this work when multiple indices have been modified
```
[the remark this answers is not in the document]
Comment (uvafan) [reply textattack/constraints/semantics/sentence_encoders/sentence_encoder.py:96]: Yes eventually we want to check all to work for WordSwap transformations which swap multiple words. As we discussed transformations like these could be useful in augmentation contexts.

File: textattack/constraints/grammaticality/language_models/google_language_model/google_language_model.py (hunk 7 above)
Comment (uvafan) [reply textattack/constraints/grammaticality/language_models/google_language_model/google_language_model.py:103]: It's a list to stay consistent with the call_many method it overrides. Good point, I'll change it to the list comprehension way.

**Pull request merged**